// g711/dist/module.js
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $271eead81bbc4ac4$exports = {};

$parcel$export($271eead81bbc4ac4$exports, "alawFromPCM", () => $271eead81bbc4ac4$export$55368bfb610ada6a);
$parcel$export($271eead81bbc4ac4$exports, "alawToPCM", () => $271eead81bbc4ac4$export$932587e2e5fe36a);
/**
 * G711U(u-law) - PCM converter
 */ const $271eead81bbc4ac4$var$SIGN_BIT = 0x80;
const $271eead81bbc4ac4$var$QUANT_MASK = 0x0f;
const $271eead81bbc4ac4$var$SEG_SHIFT = 0x04;
const $271eead81bbc4ac4$var$SEG_MASK = 0x70;
// const seg_end = new Uint16Array([0x1F, 0x3F, 0x7F, 0xFF, 0x1FF, 0x3FF, 0x7FF, 0xFFF]);
const $271eead81bbc4ac4$var$seg_end = new Uint16Array([
    0xFF,
    0x1FF,
    0x3FF,
    0x7FF,
    0xFFF,
    0x1FFF,
    0x3FFF,
    0x7FFF
]);
// const seg_uend = new Uint16Array([0x3F, 0x7F, 0xFF, 0x1FF, 0x3FF, 0x7FF, 0xFFF, 0x1FFF]);
function $271eead81bbc4ac4$var$search(val, table, size) {
    for(let i = 0; i < size; i++){
        if (val <= table[i]) return i;
    }
    return size;
}
/**
  *
  * @param {Int16Array} data
  * @returns
  */ function $271eead81bbc4ac4$var$linear2alaw(pcm_val) /* 2's complement (16-bit range) */ {
    let mask;
    let seg;
    let aval;
    if (pcm_val >= 0) mask = 0xD5; /* sign (7th) bit = 1 */
    else {
        mask = 0x55; /* sign bit = 0 */
        pcm_val = -pcm_val - 8;
    }
    /* Convert the scaled magnitude to segment number. */ seg = $271eead81bbc4ac4$var$search(pcm_val, $271eead81bbc4ac4$var$seg_end, 8);
    /* Combine the sign, segment, and quantization bits. */ if (seg >= 8) /* out of range, return maximum value. */ return 0x7F ^ mask;
    else {
        aval = seg << $271eead81bbc4ac4$var$SEG_SHIFT;
        if (seg < 2) aval |= pcm_val >> 4 & $271eead81bbc4ac4$var$QUANT_MASK;
        else aval |= pcm_val >> seg + 3 & $271eead81bbc4ac4$var$QUANT_MASK;
        return aval ^ mask;
    }
}
function $271eead81bbc4ac4$export$55368bfb610ada6a(data) {
    let dest = new Uint8Array(data.length);
    for(let i = 0, k = 0; i < data.length; i++)dest[k++] = $271eead81bbc4ac4$var$linear2alaw(data[i]);
    return dest;
}
function $271eead81bbc4ac4$var$alaw2linear(a_val) {
    let t;
    let seg;
    a_val ^= 0x55;
    t = (a_val & $271eead81bbc4ac4$var$QUANT_MASK) << 4;
    seg = (a_val & $271eead81bbc4ac4$var$SEG_MASK) >> $271eead81bbc4ac4$var$SEG_SHIFT;
    switch(seg){
        case 0:
            t += 8;
            break;
        case 1:
            t += 0x108;
            break;
        default:
            t += 0x108;
            t <<= seg - 1;
    }
    return a_val & $271eead81bbc4ac4$var$SIGN_BIT ? t : -t;
}
function $271eead81bbc4ac4$export$932587e2e5fe36a(data, bit = 16) {
    let typedArray = bit === 16 ? Int16Array : Int8Array;
    let dest = new typedArray(data.length);
    for(let i = 0, k = 0, len = data.length; i < len; i++)dest[k++] = $271eead81bbc4ac4$var$alaw2linear(data[i] & 0xff);
    return dest;
}


var $2daa776076d7c398$exports = {};

$parcel$export($2daa776076d7c398$exports, "ulaw_to_linear", () => $2daa776076d7c398$export$d374778eaff95fc7);
$parcel$export($2daa776076d7c398$exports, "ulawToPCM", () => $2daa776076d7c398$export$b1e3df51c02f2b67);
$parcel$export($2daa776076d7c398$exports, "ulawFromPCM", () => $2daa776076d7c398$export$14fab4ef5652807f);
/**
 * G711U(u-law) - PCM converter
 */ const $2daa776076d7c398$var$SIGN_BIT = 0x80;
const $2daa776076d7c398$var$QUANT_MASK = 0x0f;
const $2daa776076d7c398$var$SEG_SHIFT = 0x04;
const $2daa776076d7c398$var$SEG_MASK = 0x70;
const $2daa776076d7c398$var$BIAS = 0x84;
// const seg_end = new Uint16Array([0x1F, 0x3F, 0x7F, 0xFF, 0x1FF, 0x3FF, 0x7FF, 0xFFF]);
const $2daa776076d7c398$var$seg_end = new Uint16Array([
    0xFF,
    0x1FF,
    0x3FF,
    0x7FF,
    0xFFF,
    0x1FFF,
    0x3FFF,
    0x7FFF
]);
const $2daa776076d7c398$export$d374778eaff95fc7 = new Array(256);
// 初始化ulaw表
for(let i = 0; i < 256; i++)$2daa776076d7c398$export$d374778eaff95fc7[i] = $2daa776076d7c398$var$ulaw2linear(i);
function $2daa776076d7c398$var$ulaw2linear(u_val) {
    let t;
    u_val = ~u_val;
    t = ((u_val & $2daa776076d7c398$var$QUANT_MASK) << 3) + $2daa776076d7c398$var$BIAS;
    t <<= (u_val & $2daa776076d7c398$var$SEG_MASK) >>> $2daa776076d7c398$var$SEG_SHIFT;
    return (u_val & $2daa776076d7c398$var$SIGN_BIT) > 0 ? $2daa776076d7c398$var$BIAS - t : t - $2daa776076d7c398$var$BIAS;
}
function $2daa776076d7c398$var$linear2ulaw(pcm_val) {
    let mask;
    let seg;
    let uval;
    /* Get the sign and the magnitude of the value. */ if (pcm_val < 0) {
        pcm_val = $2daa776076d7c398$var$BIAS - pcm_val;
        mask = 0x7F;
    } else {
        pcm_val += $2daa776076d7c398$var$BIAS;
        mask = 0xFF;
    }
    /* Convert the scaled magnitude to segment number. */ seg = $2daa776076d7c398$var$search(pcm_val, $2daa776076d7c398$var$seg_end, 8);
    /*
      * Combine the sign, segment, quantization bits;
      * and complement the code word.
      */ if (seg >= 8) /* out of range, return maximum value. */ return 0x7F ^ mask;
    else {
        uval = seg << 4 | pcm_val >> seg + 3 & 0xF;
        return uval ^ mask;
    }
}
function $2daa776076d7c398$var$search(val, table, size) {
    for(let i1 = 0; i1 < size; i1++){
        if (val <= table[i1]) return i1;
    }
    return size;
}
function $2daa776076d7c398$var$toShortArray(src) {
    let dst = new Int16Array(src.length / 2);
    for(let i2 = 0, k = 0; i2 < src.length;)dst[k++] = src[i2++] & 0xff | (src[i2++] & 0xff) << 8;
    return dst;
}
function $2daa776076d7c398$export$b1e3df51c02f2b67(data, bit = 16) {
    let typedArray = bit === 16 ? Int16Array : Int8Array;
    let dest = new typedArray(data.length);
    for(let i3 = 0, k = 0, len = data.length; i3 < len; i3++)dest[k++] = $2daa776076d7c398$export$d374778eaff95fc7[data[i3] & 0xff];
    return dest;
}
function $2daa776076d7c398$export$14fab4ef5652807f(data) {
    let dest = new Uint8Array(data.length);
    for(let i4 = 0, k = 0; i4 < data.length; i4++)dest[k++] = $2daa776076d7c398$var$linear2ulaw(data[i4]);
    return dest;
}


var $2e6f12cfe07f95cc$exports = {};

$parcel$export($2e6f12cfe07f95cc$exports, "encodeWAV", () => $2e6f12cfe07f95cc$export$ccb387d635f07e97);
function $2e6f12cfe07f95cc$export$ccb387d635f07e97(bytes, sampleRate = 44100, numChannels = 2, sampleBits = 16, littleEndian = true) {
    let buffer = new ArrayBuffer(44 + bytes.byteLength), data = new DataView(buffer), offset = 0;
    // 资源交换文件标识符
    $2e6f12cfe07f95cc$var$writeString(data, offset, "RIFF");
    offset += 4;
    // 下个地址开始到文件尾总字节数
    data.setUint32(offset, 44 + bytes.byteLength, littleEndian);
    offset += 4;
    // WAV文件标志
    $2e6f12cfe07f95cc$var$writeString(data, offset, "WAVE");
    offset += 4;
    // 波形格式标志
    $2e6f12cfe07f95cc$var$writeString(data, offset, "fmt ");
    offset += 4;
    // 过滤字节,一般为 0x10 = 16
    data.setUint32(offset, 16, littleEndian);
    offset += 4;
    // 格式类别 (PCM形式采样数据)
    data.setUint16(offset, 1, littleEndian);
    offset += 2;
    // 声道数
    data.setUint16(offset, numChannels, littleEndian);
    offset += 2;
    // 采样率,每秒样本数,表示每个通道的播放速度
    data.setUint32(offset, sampleRate, littleEndian);
    offset += 4;
    // 波形数据传输率 (每秒平均字节数) 声道数 × 采样频率 × 采样位数 / 8
    data.setUint32(offset, numChannels * sampleRate * (sampleBits / 8), littleEndian);
    offset += 4;
    // 快数据调整数 采样一次占用字节数 声道数 × 采样位数 / 8
    data.setUint16(offset, numChannels * (sampleBits / 8), littleEndian);
    offset += 2;
    // 采样位数
    data.setUint16(offset, sampleBits, littleEndian);
    offset += 2;
    // 数据标识符
    $2e6f12cfe07f95cc$var$writeString(data, offset, "data");
    offset += 4;
    // 采样数据总数,即数据总大小-44
    data.setUint32(offset, bytes.byteLength, littleEndian);
    offset += 4;
    // 给wav头增加pcm体
    for(let i = 0; i < bytes.byteLength;){
        data.setUint8(offset, bytes.getUint8(i));
        offset++;
        i++;
    }
    return buffer;
}
function $2e6f12cfe07f95cc$var$writeString(view, offset, string) {
    for(let i = 0, len = string.length; i < len; i++)view.setUint8(offset + i, string.charCodeAt(i));
}




export {$271eead81bbc4ac4$export$55368bfb610ada6a as alawFromPCM, $271eead81bbc4ac4$export$932587e2e5fe36a as alawToPCM, $2daa776076d7c398$export$d374778eaff95fc7 as ulaw_to_linear, $2daa776076d7c398$export$b1e3df51c02f2b67 as ulawToPCM, $2daa776076d7c398$export$14fab4ef5652807f as ulawFromPCM, $2e6f12cfe07f95cc$export$ccb387d635f07e97 as encodeWAV};
